src/plugins/vis_builder/public/application/utils/get_persisted_agg_params.ts

```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import { CreateAggConfigParams, propFilter } from '../../../../data/common';
+import { Schema } from '../../../../vis_default_editor/public';
+
+const filterByType = propFilter('type');
+
+export const getPersistedAggParams = (
+  aggConfigParams: CreateAggConfigParams[],
+  oldVisSchemas: Schema[] = [],
+  newVisSchemas: Schema[] = []
+): CreateAggConfigParams[] => {
+  const updatedAggConfigParams: CreateAggConfigParams[] = [];
+  const newVisSchemaCounts: Record<string, number> = newVisSchemas.reduce((acc, schema: Schema) => {
+    acc[schema.name] = schema.max;
+    return acc;
+  }, {});
+
+  // For each aggConfigParam, check if a compatible schema exists in the new visualization type
+  aggConfigParams.forEach((aggConfigParam) => {
+    const currentSchema = oldVisSchemas.find((schema: Schema) => {
+      return schema.name === aggConfigParam.schema;
+    });
+
+    // see if a matching schma exists in the new visualization type
+    const matchingSchema = newVisSchemas.find((schema: Schema) => {
+      return schema.name === aggConfigParam.schema;
+    });
+
+    // if the matching schema is same as the current schema, add the aggConfigParam to the updatedAggConfigParams
+    if (
+      isSchemaEqual(matchingSchema, currentSchema) &&
+      newVisSchemaCounts[matchingSchema!.name] > 0
+    ) {
+      updatedAggConfigParams.push(aggConfigParam);
+      newVisSchemaCounts[matchingSchema!.name] -= 1;
+      return;
+    }
+
+    // if a matching schema does not exist, check if a compatible schema exists
+    for (const schema of newVisSchemas) {
+      // Check if the schema group is the same
+      if (schema.group !== currentSchema!.group) continue;
+
+      const compatibleSchema = filterByType([aggConfigParam], schema.aggFilter).length !== 0;
```

Comment:
This was the main reason for the change. So each schema today can define a list of agg types it supports. Its not a simple lift and it can sometimes be something like `!filter`, In this case we want to make sure that when we select a new schema type to map the agg to, the agg type for the agg config is allowed by the new schema.

Reply:
Thanks, make sense